File: apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/ProviderHttpTest.kt
package com.agentic.wallet.agent.provider

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class ProviderHttpTest {
    @Test
    fun mapsHttpStatus() {
        assertNull(ProviderHttp.mapHttpStatusToErrorCode(200))
        assertEquals(ProviderErrorCodes.AUTH, ProviderHttp.mapHttpStatusToErrorCode(401))
        assertEquals(ProviderErrorCodes.AUTH, ProviderHttp.mapHttpStatusToErrorCode(403))
        assertEquals(ProviderErrorCodes.RATE_LIMITED, ProviderHttp.mapHttpStatusToErrorCode(429))
        assertEquals(ProviderErrorCodes.TIMEOUT, ProviderHttp.mapHttpStatusToErrorCode(408))
        assertEquals(ProviderErrorCodes.TIMEOUT, ProviderHttp.mapHttpStatusToErrorCode(504))
        assertEquals(ProviderErrorCodes.UPSTREAM, ProviderHttp.mapHttpStatusToErrorCode(500))
        assertEquals(ProviderErrorCodes.UPSTREAM, ProviderHttp.mapHttpStatusToErrorCode(503))
        assertEquals(ProviderErrorCodes.INVALID_RESPONSE, ProviderHttp.mapHttpStatusToErrorCode(418))
    }

    @Test
    fun normalizeBaseUrlAppendsV1ForOpenAi() {
        assertEquals("https://api.openai.com/v1", ProviderHttp.normalizeBaseUrl("https://api.openai.com", "openai-compatible"))
        assertEquals("https://api.openai.com/v1", ProviderHttp.normalizeBaseUrl("https://api.openai.com/", "openai-compatible"))
        assertEquals("https://api.openai.com/v1", ProviderHttp.normalizeBaseUrl("https://api.openai.com/v1", "openai-compatible"))
        assertEquals("https://generativelanguage.googleapis.com/v1beta/openai", ProviderHttp.normalizeBaseUrl("https://generativelanguage.googleapis.com/v1beta/openai", "openai-compatible"))
    }

    @Test
    fun normalizeBaseUrlAppendsV1ForAnthropic() {
        assertEquals("https://api.anthropic.com/v1", ProviderHttp.normalizeBaseUrl("https://api.anthropic.com", "anthropic"))
        assertEquals("https://api.anthropic.com/v1", ProviderHttp.normalizeBaseUrl("https://api.anthropic.com/", "anthropic"))
        assertEquals("https://api.anthropic.com/v2", ProviderHttp.normalizeBaseUrl("https://api.anthropic.com/v2", "anthropic"))
    }

    @Test
    fun normalizeBaseUrlFallsBackToPresetWhenBlank() {
        assertEquals("https://api.openai.com/v1", ProviderHttp.normalizeBaseUrl("", "openai-compatible"))
        assertEquals("https://api.openai.com/v1", ProviderHttp.normalizeBaseUrl(null, "openai-compatible"))
        assertEquals("https://api.anthropic.com/v1", ProviderHttp.normalizeBaseUrl("", "anthropic"))
    }

    @Test
    fun isDefaultTemperatureOnlyModelDetectsGpt5AndOSeries() {
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("gpt-5"))
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("gpt-5-turbo"))
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("provider/gpt-5"))
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("o1-preview"))
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("o3-mini"))
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("provider/o1"))
        assertTrue(ProviderHttp.isDefaultTemperatureOnlyModel("provider/o4-mini"))
        assertFalse(ProviderHttp.isDefaultTemperatureOnlyModel("gpt-4o"))
        assertFalse(ProviderHttp.isDefaultTemperatureOnlyModel("claude-opus-4-5"))
        assertFalse(ProviderHttp.isDefaultTemperatureOnlyModel("gemini-1.5-pro"))
        assertFalse(ProviderHttp.isDefaultTemperatureOnlyModel(""))
    }

    @Test
    fun assertApiKeyHeaderSafeAcceptsAscii() {
        ProviderHttp.assertApiKeyHeaderSafe("sk-ABCDEF1234567890")
        ProviderHttp.assertApiKeyHeaderSafe("xoxp-1234567890abcdef")
    }

    @Test
    fun assertApiKeyHeaderSafeRejectsControlChars() {
        val ex = assertThrows(ProviderHttpException::class.java) {
            ProviderHttp.assertApiKeyHeaderSafe("sk-\nABCDEF")
        }
        assertEquals(ProviderErrorCodes.INVALID_CONFIG, ex.code)
    }

    @Test
    fun assertApiKeyHeaderSafeRejectsNonAscii() {
        val ex = assertThrows(ProviderHttpException::class.java) {
            ProviderHttp.assertApiKeyHeaderSafe("sk-ABCéDEF")
        }
        assertEquals(ProviderErrorCodes.INVALID_CONFIG, ex.code)
    }

    @Test
    fun assertApiKeyHeaderSafeRejectsEmpty() {
        val ex = assertThrows(ProviderHttpException::class.java) {
            ProviderHttp.assertApiKeyHeaderSafe("")
        }
        assertEquals(ProviderErrorCodes.INVALID_CONFIG, ex.code)
        assertTrue(
            "Empty-key message should be clear, got: ${ex.message}",
            ex.message.contains("empty", ignoreCase = true),
        )
    }

    @Test
    fun composeErrorMessageWithJsonErrorMessage() {
        val composed = ProviderHttp.composeErrorMessage(
            401,
            """{"error":{"message":"key invalid"}}""",
        )
        assertTrue("Should start with upstream message: $composed", composed.startsWith("key invalid."))
        assertTrue("Should append 401 explanation: $composed", composed.contains("belongs to this provider"))
    }

    @Test
    fun composeErrorMessageWithRawErrorString() {
        val composed = ProviderHttp.composeErrorMessage(
            429,
            """{"error":"limit reached"}""",
        )
        assertTrue("Should include upstream msg: $composed", composed.contains("limit reached"))
        assertTrue("Should append 429 explanation: $composed", composed.contains("too many requests"))
    }

    @Test
    fun composeErrorMessageWithBlankBody() {
        val composed = ProviderHttp.composeErrorMessage(503, "")
        assertTrue(
            "Should fall back to HTTP message: $composed",
            composed.startsWith("AI provider returned HTTP 503."),
        )
        assertTrue("Should append 503 explanation: $composed", composed.contains("temporarily unavailable"))
    }

    @Test
    fun composeErrorMessageUnknownStatusOmitsExplanation() {
        val composed = ProviderHttp.composeErrorMessage(200, """{"error":{"message":"weird"}}""")
        // 200 has no explanation; output is just the base message, no trailing whitespace.
        assertEquals("weird", composed)
    }

    @Test
    fun composeErrorMessageNoDoublePeriods() {
        val composed = ProviderHttp.composeErrorMessage(
            429,
            """{"error":{"message":"already a sentence."}}""",
        )
        assertFalse("Must not double-period: $composed", composed.contains(".."))
    }

    @Test
    fun composeErrorMessageAddsSeparatorWhenBaseHasNoTerminal() {
        val composed = ProviderHttp.composeErrorMessage(
            500,
            """{"error":{"message":"backend exploded"}}""",
        )
        assertTrue("Should insert a period: $composed", composed.startsWith("backend exploded. "))
    }

    @Test
    fun composeErrorMessageHandlesMalformedBody() {
        val composed = ProviderHttp.composeErrorMessage(500, "not even json")
        assertTrue(composed.startsWith("AI provider returned HTTP 500."))
    }
}

File: apps/browser-demo/src/deviceAgent/__tests__/providerHttp.test.ts
// Ported from
// apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/ProviderHttpTest.kt.
// Pins URL normalization, HTTP status mapping, the gpt-5 / o-series temperature
// detection, the API-key header guard, and composeErrorMessage punctuation logic.

import { describe, expect, it } from 'vitest';

import { ProviderHttpError } from '../provider/errorCodes.js';
import {
  assertApiKeyHeaderSafe,
  composeErrorMessage,
  isDefaultTemperatureOnlyModel,
  mapHttpStatusToErrorCode,
  normalizeBaseUrl,
} from '../provider/providerHttp.js';

describe('mapHttpStatusToErrorCode', () => {
  it('maps statuses to provider_* codes', () => {
    expect(mapHttpStatusToErrorCode(200)).toBeNull();
    expect(mapHttpStatusToErrorCode(204)).toBeNull();
    expect(mapHttpStatusToErrorCode(401)).toBe('provider_auth');
    expect(mapHttpStatusToErrorCode(403)).toBe('provider_auth');
    expect(mapHttpStatusToErrorCode(429)).toBe('provider_rate_limited');
    expect(mapHttpStatusToErrorCode(408)).toBe('provider_timeout');
    expect(mapHttpStatusToErrorCode(504)).toBe('provider_timeout');
    expect(mapHttpStatusToErrorCode(500)).toBe('provider_upstream');
    expect(mapHttpStatusToErrorCode(503)).toBe('provider_upstream');
    expect(mapHttpStatusToErrorCode(418)).toBe('provider_invalid_response');
  });
});

describe('normalizeBaseUrl — openai-compatible', () => {
  it('appends /v1 to bare host', () => {
    expect(normalizeBaseUrl('https://api.openai.com', 'openai-compatible')).toBe(
      'https://api.openai.com/v1',
    );
  });

  it('strips trailing slash and appends /v1', () => {
    expect(normalizeBaseUrl('https://api.openai.com/', 'openai-compatible')).toBe(
      'https://api.openai.com/v1',
    );
  });

  it('leaves an already-versioned URL alone', () => {
    expect(normalizeBaseUrl('https://api.openai.com/v1', 'openai-compatible')).toBe(
      'https://api.openai.com/v1',
    );
  });

  it('leaves the Gemini /v1beta/openai suffix alone', () => {
    expect(
      normalizeBaseUrl('https://generativelanguage.googleapis.com/v1beta/openai', 'openai-compatible'),
    ).toBe('https://generativelanguage.googleapis.com/v1beta/openai');
  });

  it('falls back to the OpenAI preset for blank / null / undefined input', () => {
    expect(normalizeBaseUrl('', 'openai-compatible')).toBe('https://api.openai.com/v1');
    expect(normalizeBaseUrl(null, 'openai-compatible')).toBe('https://api.openai.com/v1');
    expect(normalizeBaseUrl(undefined, 'openai-compatible')).toBe('https://api.openai.com/v1');
  });
});

describe('normalizeBaseUrl — anthropic', () => {
  it('appends /v1 to bare host', () => {
    expect(normalizeBaseUrl('https://api.anthropic.com', 'anthropic')).toBe('https://api.anthropic.com/v1');
  });

  it('strips trailing slash and appends /v1', () => {
    expect(normalizeBaseUrl('https://api.anthropic.com/', 'anthropic')).toBe('https://api.anthropic.com/v1');
  });

  it('leaves a /v2 (non-v1) version alone', () => {
    expect(normalizeBaseUrl('https://api.anthropic.com/v2', 'anthropic')).toBe('https://api.anthropic.com/v2');
  });

  it('falls back to the Anthropic preset for blank input', () => {
    expect(normalizeBaseUrl('', 'anthropic')).toBe('https://api.anthropic.com/v1');
  });
});

describe('isDefaultTemperatureOnlyModel', () => {
  it('returns true for gpt-5 family', () => {
    expect(isDefaultTemperatureOnlyModel('gpt-5')).toBe(true);
    expect(isDefaultTemperatureOnlyModel('gpt-5-turbo')).toBe(true);
    expect(isDefaultTemperatureOnlyModel('provider/gpt-5')).toBe(true);
  });

  it('returns true for o-series', () => {
    expect(isDefaultTemperatureOnlyModel('o1-preview')).toBe(true);
    expect(isDefaultTemperatureOnlyModel('o3-mini')).toBe(true);
    expect(isDefaultTemperatureOnlyModel('provider/o1')).toBe(true);
    expect(isDefaultTemperatureOnlyModel('provider/o4-mini')).toBe(true);
  });

  it('returns false for ordinary models and blank input', () => {
    expect(isDefaultTemperatureOnlyModel('gpt-4o')).toBe(false);
    expect(isDefaultTemperatureOnlyModel('claude-opus-4-5')).toBe(false);
    expect(isDefaultTemperatureOnlyModel('gemini-1.5-pro')).toBe(false);
    expect(isDefaultTemperatureOnlyModel('')).toBe(false);
  });
});

describe('assertApiKeyHeaderSafe', () => {
  it('accepts ASCII printable keys', () => {
    expect(() => assertApiKeyHeaderSafe('sk-ABCDEF1234567890')).not.toThrow();
    expect(() => assertApiKeyHeaderSafe('xoxp-1234567890abcdef')).not.toThrow();
  });

  it('rejects keys with a control character', () => {
    let captured: unknown = null;
    try {
      assertApiKeyHeaderSafe('sk-\nABCDEF');
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    expect((captured as ProviderHttpError).code).toBe('provider_invalid_config');
  });

  it('rejects keys with a non-ASCII character', () => {
    let captured: unknown = null;
    try {
      assertApiKeyHeaderSafe('sk-ABCéDEF');
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    expect((captured as ProviderHttpError).code).toBe('provider_invalid_config');
  });

  it('rejects an empty key with an "empty" message', () => {
    let captured: unknown = null;
    try {
      assertApiKeyHeaderSafe('');
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    const err = captured as ProviderHttpError;
    expect(err.code).toBe('provider_invalid_config');
    expect(err.message.toLowerCase().includes('empty')).toBe(true);
  });
});

describe('composeErrorMessage', () => {
  it('formats {error:{message}} JSON with appended explanation', () => {
    const composed = composeErrorMessage(401, '{"error":{"message":"key invalid"}}');
    expect(composed.startsWith('key invalid.')).toBe(true);
    expect(composed.includes('belongs to this provider')).toBe(true);
  });

  it('accepts a raw string in the error field', () => {
    const composed = composeErrorMessage(429, '{"error":"limit reached"}');
    expect(composed.includes('limit reached')).toBe(true);
    expect(composed.includes('too many requests')).toBe(true);
  });

  it('falls back to "AI provider returned HTTP <status>." when body is blank', () => {
    const composed = composeErrorMessage(503, '');
    expect(composed.startsWith('AI provider returned HTTP 503.')).toBe(true);
    expect(composed.includes('temporarily unavailable')).toBe(true);
  });

  it('omits the explanation for status with no entry (200)', () => {
    expect(composeErrorMessage(200, '{"error":{"message":"weird"}}')).toBe('weird');
  });

  it('avoids double periods when the base message ends with a period', () => {
    const composed = composeErrorMessage(429, '{"error":{"message":"already a sentence."}}');
    expect(composed.includes('..')).toBe(false);
  });

  it('inserts ". " when the base message lacks terminal punctuation', () => {
    const composed = composeErrorMessage(500, '{"error":{"message":"backend exploded"}}');
    expect(composed.startsWith('backend exploded. ')).toBe(true);
  });

  it('handles malformed body by falling back to the HTTP status message', () => {
    const composed = composeErrorMessage(500, 'not even json');
    expect(composed.startsWith('AI provider returned HTTP 500.')).toBe(true);
  });
});
